### DISCUSSION TITLE
External kinematics?

### DISCUSSION THREAD

**Discussion by alexmkrolak** [visose/Robots#163] [Q&A] [answered]
Are there any examples of doing something similar to external kinematics for a vertical positioner/turntable, for example for robotic milling of a bust like in [this video](https://www.youtube.com/watch?v=LImgOPV1tgk&t=195s) using KUKA prc?

**Comment by visose** [comment, marked as answer]
The video above is set to private, but I can guess what you are after. I added an example with a positioner [here](https://github.com/visose/Robots/tree/master/samples/Robots.Samples.Grasshopper).

It's a minimal example to understand how to align the positioner to follow the toolpath. It's not a complete milling example, you would need to add proper approaches, retractions and so on.

If the shape of the object has undercuts or is more complex, it might be better to generate the toolpath using CNC software and then bring it to Grasshopper. You can use the [Extensions](https://github.com/visose/Extensions) plugin to import from RhinoCAM and PowerMill to Robots.

**Reply by alexmkrolak** [reply]
Thank you, I've now figured this out using your example!

In terms of calibrating the turntable, there is a line in the robot .xml file saying:

<Positioner model="KP1-V500" manufacturer="KUKA" payload="500">
       <Base x="1382.72" y="665.46" z="0.00" q1="0.992071" q2="-0.000807" q3="-0.000174" q4="0.125679"/>

Could you clarify what the q1-4 values are referring to? I can assume that its vector for the the rotational axis of the positioner but not sure why there are 4 values?

**Reply by visose** [reply]
These are quaternions. In the xml, a coordinate system (plane) is defined by a point (for position) and a quaternion (for 3d rotation). So 7 numbers: X,Y,Z,Q1,Q2,Q3,Q4.
The KUKA controller uses a different representation, it uses a point + Euler angles (X,Y,Z, A,B,C). You can convert the values from the controller to the quaternion representation required by the xml by using the "Get Plane" and "From Plane" components.

**Reply by alexmkrolak** [reply]
If "From Plane" is only giving me 3 values for quaternions, is q4 just zero then?

**Reply by alexmkrolak** [reply]
![image](https://user-images.githubusercontent.com/69169386/174813039-b80e7e52-7776-4483-8118-f112baaa1dee.png)

You can see here the plane I'm inputting into the "From Plane" block and what the result is. Ignore the Z-height

[foam milling.zip](https://github.com/visose/Robots/files/8949412/foam.milling.zip)

**Reply by visose** [reply]
If you connect a robot to the R input, it will use the convention of that robot, so if you are connecting a KUKA robot, it will return XYZABC, if you don't connect anything it will return quaternions. If you hover over the component you will find this explanation on the tooltip.

**Reply by visose** [reply]
<img width="584" alt="image" src="https://user-images.githubusercontent.com/5129614/174818903-1a451d5b-715b-4b4f-b4c8-4a35093a8182.png">
It should look like this

**Reply by visose** [reply]
I see in your file that you are inputting in Grasshopper some sample coordinates you are using to calibrate the base and then using "Plane Fit" component to find a plane. This will not match with the calibrated plane in the controller.

You should calibrate the positioner using the KRC4 pendant, once calibrated it will give you the X,Y,Z,A,B,C values of the positioner. Only input these values.

**Reply by visose** [reply]
One way to find these values in the controller is by opening the file `R1/System/$config.dat`, then scroll to the `MACHINE_DEF` array, the first item will be the robot, probably with the base at 0,0,0,0,0,0 if you never changed it. The second item is probably your positioner, unless you have more external mechanisms.

**Reply by alexmkrolak** [reply]
> If you hover over the component you will find this explanation on the tooltip.

Ah yes, apologies, I missed that.

> You should calibrate the positioner using the KRC4 pendant, once calibrated it will give you the X,Y,Z,A,B,C values of the positioner.

So I've also been having issues with this step since I believe I am doing the calibration correctly but then it tells me that i have an error much larger (around 80-100mm) than the maximum error allowed (5mm). I took a [video](https://www.youtube.com/watch?v=TJX9cj7bfwA) of me performing the calibration - does it look like I'm doing anything stupidly long?

> One way to find these values in the controller is by opening the file `R1/System/$config.dat`

Am I right in thinking that changing the positioner base in the *.xml file only changes the simulation in Grasshopper, but the actual KRL code only ever takes the root value from the $config.dat file on the KRC4? So therefore the only way to calibrate the simulation to the physical robot movements is to change the root value in the $config.dat and make sure that the *.xml values match the $config.dat values? There's no way to jog the root values through Robots/Grasshopper?

**Reply by visose** [reply]
It's difficult to troubleshoot from a video, but I live less than a 10 min walk away from that robot, email me if you have budget for on-site support 😄

**Reply by visose** [reply]
The xml contains data from the controller, which is needed as you say make a proper simulation of your robot cell. This is stored in the xml because its not frequently changed. It's not part of the generated program because the controller already contains this data (so it's important that they match). Data that you want to input into the program and the controller doesn't know about goes in the Grasshopper file.
